File: flutter/weft_fintech/lib/src/mdp1_wire.dart
// mdp1_wire.dart — MDP1 aggregated book snapshot wire, Dart side.
//
// Mirrors packages/fintech/src/mdp1.js byte-for-byte (Pillar 6 / Law 4):
//   304-byte little-endian record:
//     0   magic "MDP1" | 4 version u16 | 6 flags u16
//     8   seq u64 | 16 last_ts_ns u64 | 24 best_bid u32 | 28 best_ask u32
//     32  bid_levels[10] x12 {price,size,orders} | 152 ask_levels[10] x12
//     272 msg_count u64 | 280 trade_count u64 | 288 last_match u64
//     296 crc32 u32 (poly 0xEDB88320, init/final 0xFFFFFFFF, over [0,296))
//     300 reserved u32
//
// All multi-byte reads use ByteData with EXPLICIT Endian.little (Law 2).
// u64 fields compose as lo + hi * 4294967296 — exact Dart ints, never
// floats. The flyweight binds to the CALLER's bytes: book state never
// crosses into widget state (Law 4 / W4-06).
//
// Audit contract: constants below are `const int <name> = <value>;` so
// test/static_audit.mjs can diff them against the TS reference.

// ignore_for_file: constant_identifier_names

library;

import 'dart:typed_data';

const int MDP1_SIZE = 304;
const int MDP1_VERSION = 1;
const int MDP1_TOP_LEVELS = 10;

const int F_BOOK_VALID = 1 << 0;
const int F_CROSSED = 1 << 1;
const int F_LOCKED = 1 << 2;

const int MDP1_OFF_VERSION = 4;
const int MDP1_OFF_FLAGS = 6;
const int MDP1_OFF_SEQ = 8;
const int MDP1_OFF_LAST_TS = 16;
const int MDP1_OFF_BEST_BID = 24;
const int MDP1_OFF_BEST_ASK = 28;
const int MDP1_OFF_BIDS = 32;
const int MDP1_OFF_ASKS = 152;
const int MDP1_OFF_MSG_COUNT = 272;
const int MDP1_OFF_TRADE_COUNT = 280;
const int MDP1_OFF_LAST_MATCH = 288;
const int MDP1_OFF_CRC = 296;
const int MDP1_CRC_END = 296; // CRC covers [0, 296)

const List<int> MDP1_MAGIC_BYTES = <int>[0x4D, 0x44, 0x50, 0x31]; // "MDP1"
const int MDP1_MAGIC = 0x3150444d; // little-endian u32 read of "MDP1"

// Typed failure codes — Law 4 taxonomy (TS decision order).
const int MDP1_OK = 0;
const int MDP1_E_SHORT = 1; // < 304 bytes
const int MDP1_E_MAGIC = 2;
const int MDP1_E_VERSION = 3;
const int MDP1_E_CRC = 4;

Uint32List? _crcTable;

/// CRC-32/ISO-HDLC table (poly 0xEDB88320), built once per isolate.
Uint32List crcTable() {
  final t = _crcTable;
  if (t != null) return t;
  final built = Uint32List(256);
  for (int n = 0; n < 256; n++) {
    int c = n;
    for (int k = 0; k < 8; k++) {
      c = (c & 1) != 0 ? (0xedb88320 ^ (c >>> 1)) : (c >>> 1);
    }
    built[n] = c & 0xFFFFFFFF;
  }
  _crcTable = built;
  return built;
}

/// CRC-32 over bytes [start, end) of `u8` (defaults to the MDP1 header
/// region [0, 296)). Table-driven, allocation-free per call.
int mdp1Crc32(Uint8List u8, [int start = 0, int end = MDP1_CRC_END]) {
  final table = _crcTable ?? crcTable();
  int crc = 0xFFFFFFFF;
  for (int i = start; i < end; i++) {
    crc = (crc >>> 8) ^ table[(crc ^ u8[i]) & 0xFF];
  }
  return (crc ^ 0xFFFFFFFF) & 0xFFFFFFFF;
}

/// Reusable flyweight over one MDP1 record. `retarget` re-points it at new
/// bytes — the widget owns ONE snapshot for the whole feed lifetime.
class Mdp1Snapshot {
  ByteData _dv;
  Uint8List _u8;

  Mdp1Snapshot(Uint8List buffer)
      : _dv = ByteData.sublistView(buffer),
        _u8 = buffer;

  /// Re-point this flyweight at a fresh record (zero allocation beyond the
  /// single sublist view — performed once per delivered frame, never per
  /// level or per paint).
  void retarget(Uint8List buffer) {
    _u8 = buffer;
    _dv = ByteData.sublistView(buffer);
  }

  /// Fail-closed structural check: returns MDP1_OK or a typed E_* code
  /// evaluated in the TS decision order (short -> magic -> version -> CRC).
  int validate() {
    if (_u8.length < MDP1_SIZE) return MDP1_E_SHORT;
    if (_dv.getUint32(0, Endian.little) != MDP1_MAGIC) return MDP1_E_MAGIC;
    if (_dv.getUint16(MDP1_OFF_VERSION, Endian.little) != MDP1_VERSION) {
      return MDP1_E_VERSION;
    }
    final stored = _dv.getUint32(MDP1_OFF_CRC, Endian.little);
    if (stored != mdp1Crc32(_u8)) return MDP1_E_CRC;
    return MDP1_OK;
  }

  int get flags => _dv.getUint16(MDP1_OFF_FLAGS, Endian.little);
  bool get bookValid => (flags & F_BOOK_VALID) != 0;
  bool get crossed => (flags & F_CROSSED) != 0;
  bool get locked => (flags & F_LOCKED) != 0;

  /// u64 seq as exact Dart int (lo + hi * 2^32).
  int get seq {
    final lo = _dv.getUint32(MDP1_OFF_SEQ, Endian.little);
    final hi = _dv.getUint32(MDP1_OFF_SEQ + 4, Endian.little);
    return hi * 4294967296 + lo;
  }

  int get lastTsNs {
    final lo = _dv.getUint32(MDP1_OFF_LAST_TS, Endian.little);
    final hi = _dv.getUint32(MDP1_OFF_LAST_TS + 4, Endian.little);
    return hi * 4294967296 + lo;
  }

  int get bestBid => _dv.getUint32(MDP1_OFF_BEST_BID, Endian.little);
  int get bestAsk => _dv.getUint32(MDP1_OFF_BEST_ASK, Endian.little);

  int bidPrice(int i) =>
      _dv.getUint32(MDP1_OFF_BIDS + i * 12, Endian.little);
  int bidSize(int i) =>
      _dv.getUint32(MDP1_OFF_BIDS + i * 12 + 4, Endian.little);
  int bidOrders(int i) =>
      _dv.getUint32(MDP1_OFF_BIDS + i * 12 + 8, Endian.little);
  int askPrice(int i) =>
      _dv.getUint32(MDP1_OFF_ASKS + i * 12, Endian.little);
  int askSize(int i) =>
      _dv.getUint32(MDP1_OFF_ASKS + i * 12 + 4, Endian.little);
  int askOrders(int i) =>
      _dv.getUint32(MDP1_OFF_ASKS + i * 12 + 8, Endian.little);

  int get msgCount => _dv.getUint32(MDP1_OFF_MSG_COUNT, Endian.little);
  int get tradeCount => _dv.getUint32(MDP1_OFF_TRADE_COUNT, Endian.little);

  int get lastMatch {
    final lo = _dv.getUint32(MDP1_OFF_LAST_MATCH, Endian.little);
    final hi = _dv.getUint32(MDP1_OFF_LAST_MATCH + 4, Endian.little);
    return hi * 4294967296 + lo;
  }

  bool get crcOk =>
      _dv.getUint32(MDP1_OFF_CRC, Endian.little) == mdp1Crc32(_u8);
}

File: apple/WeftFintech/Sources/WeftFintech/Mdp1Wire.swift
// Mdp1Wire.swift — MDP1 aggregated book snapshot wire, Swift side.
//
// Mirrors packages/fintech/src/mdp1.js byte-for-byte (Pillar 6 / Law 4).
// All multi-byte loads pass .littleEndian EXPLICITLY (Law 2). u64 fields
// compose as UInt64(lo) | UInt64(hi) << 32 — exact 64-bit math, never
// Double. The view is a FLYWEIGHT over the caller's bytes: book state
// never crosses into Swift observable state except through the model's
// edge-triggered publish (Law 4 / W4-06).

import Foundation

public enum Mdp1 {
    public static let size = 304
    public static let version: UInt16 = 1
    public static let topLevels = 10
    public static let magic: UInt32 = 0x3150_444d // "MDP1" LE u32 read

    public static let fBookValid: UInt16 = 1 << 0
    public static let fCrossed: UInt16 = 1 << 1
    public static let fLocked: UInt16 = 1 << 2

    public static let offVersion = 4
    public static let offFlags = 6
    public static let offSeq = 8
    public static let offLastTs = 16
    public static let offBestBid = 24
    public static let offBestAsk = 28
    public static let offBids = 32
    public static let offAsks = 152
    public static let offMsgCount = 272
    public static let offTradeCount = 280
    public static let offLastMatch = 288
    public static let offCrc = 296
    public static let crcEnd = 296

    // Typed failure codes — Law 4 taxonomy (TS decision order).
    public static let ok = 0
    public static let eShort = 1
    public static let eMagic = 2
    public static let eVersion = 3
    public static let eCrc = 4

    static let crcTable: [UInt32] = {
        var t = [UInt32](repeating: 0, count: 256)
        for n in 0..<256 {
            var c = UInt32(n)
            for _ in 0..<8 {
                c = (c & 1) != 0 ? (0xedb8_8320 ^ (c >> 1)) : (c >> 1)
            }
            t[n] = c
        }
        return t
    }()

    /// CRC-32/ISO-HDLC over bytes [start, end) — table-driven, no allocs.
    @inlinable
    public static func crc32(_ bytes: UnsafeRawBufferPointer,
                             _ start: Int = 0,
                             _ end: Int = Mdp1.crcEnd) -> UInt32 {
        var crc: UInt32 = 0xFFFF_FFFF
        let table = Mdp1.crcTable
        for i in start..<end {
            crc = (crc >> 8) ^ table[Int((crc ^ UInt32(bytes[i])) & 0xFF)]
        }
        return crc ^ 0xFFFF_FFFF
    }
}

/// Reusable flyweight over one MDP1 record.
public struct Mdp1View {
    public let bytes: UnsafeRawBufferPointer

    public init(_ bytes: UnsafeRawBufferPointer) {
        self.bytes = bytes
    }

    /// Fail-closed structural check in the TS decision order
    /// (short -> magic -> version -> CRC). 0 == Mdp1.ok.
    public func validate() -> Int {
        if bytes.count < Mdp1.size { return Mdp1.eShort }
        if loadU32(0) != Mdp1.magic { return Mdp1.eMagic }
        if loadU16(Mdp1.offVersion) != Mdp1.version { return Mdp1.eVersion }
        if loadU32(Mdp1.offCrc) != Mdp1.crc32(bytes) { return Mdp1.eCrc }
        return Mdp1.ok
    }

    @inlinable func loadU16(_ off: Int) -> UInt16 {
        bytes.loadUnaligned(fromByteOffset: off, as: UInt16.self).littleEndian
    }
    @inlinable func loadU32(_ off: Int) -> UInt32 {
        bytes.loadUnaligned(fromByteOffset: off, as: UInt32.self).littleEndian
    }

    public var flags: UInt16 { loadU16(Mdp1.offFlags) }
    public var bookValid: Bool { flags & Mdp1.fBookValid != 0 }
    public var crossed: Bool { flags & Mdp1.fCrossed != 0 }
    public var locked: Bool { flags & Mdp1.fLocked != 0 }

    public var seq: UInt64 {
        UInt64(loadU32(Mdp1.offSeq)) |
            UInt64(loadU32(Mdp1.offSeq + 4)) << 32
    }
    public var lastTsNs: UInt64 {
        UInt64(loadU32(Mdp1.offLastTs)) |
            UInt64(loadU32(Mdp1.offLastTs + 4)) << 32
    }
    public var bestBid: UInt32 { loadU32(Mdp1.offBestBid) }
    public var bestAsk: UInt32 { loadU32(Mdp1.offBestAsk) }

    public func bidPrice(_ i: Int) -> UInt32 { loadU32(Mdp1.offBids + i * 12) }
    public func bidSize(_ i: Int) -> UInt32 { loadU32(Mdp1.offBids + i * 12 + 4) }
    public func bidOrders(_ i: Int) -> UInt32 { loadU32(Mdp1.offBids + i * 12 + 8) }
    public func askPrice(_ i: Int) -> UInt32 { loadU32(Mdp1.offAsks + i * 12) }
    public func askSize(_ i: Int) -> UInt32 { loadU32(Mdp1.offAsks + i * 12 + 4) }
    public func askOrders(_ i: Int) -> UInt32 { loadU32(Mdp1.offAsks + i * 12 + 8) }

    public var msgCount: UInt32 { loadU32(Mdp1.offMsgCount) }
    public var tradeCount: UInt32 { loadU32(Mdp1.offTradeCount) }
    public var lastMatch: UInt64 {
        UInt64(loadU32(Mdp1.offLastMatch)) |
            UInt64(loadU32(Mdp1.offLastMatch + 4)) << 32
    }

    public var crcOk: Bool { loadU32(Mdp1.offCrc) == Mdp1.crc32(bytes) }
}
